### ISSUE TITLE
to pack or not to pack ...

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#8572]
## borg 1.x segment files

borg 1.x used:
- "segment files", elsewhere also known as "pack files" to store multiple repository objects in one file.
- a "repository index" to be able to find these objects, using a mapping `object id --> (segment_name, offset_in_segment)`.
- transactions and rollback via log-like appending of operations (PUT, DEL, COMMIT) to these segment files

## borg2 status quo: objects stored separately

borg2 is much simpler:
- implemented using `borgstore` (k/v store with misc. backends)
- objects are stored separately: 1 file chunk --> 1 repo object
- objects can be directly found by their id (e.g. the id is mapped to the fs path / file name)
- no transactions, no log-like appending - but correct write order

Pros:
- simplicity
- no need for some sort of "index" (which could be corrupted or out of date)
- no segment file compaction needed, the server-side filesystem manages space allocation

Cons:
- leads to big amounts of relatively small objects transferred and stored individually in the repository
- latency and other overheads have quite a speed impact for remote repositories
- depending on the storage type / filesystem, there will be more or less storage space usage overhead due to block size, esp. for many very small objects
- dealing with lots of objects / doing lots of api calls can be expensive for some cloud storage providers

## borg2 alternative idea

- client assembles packs locally, transfers to store when the pack has reached the desired size or when there is no more data to write.
- pack files have a per-pack index appended (pointing to the objects contained in the pack), so the per-pack index can be read without reading the full pack.

Pros:
- a lot less objects in store, less api calls, less latency impact

Cons:
- more complex in general
- will need an addtl. global index mapping `object_id -> pack_id, offset_in_pack`
- space is managed clientside: compact will need to read the pack, drop unused entries and write it back to the store, fix or rebuild indexes

**Comment by dietmargoldbeck** [comment]
I would really appreciate the use of packs. Currently borg 2 is "incompatible" with most USB hard disks with SMR recording. I used a Toshiba 4TB external USB hard drive for borg2 testing and a borg check was done approx. 50% after 12 hours when i killed it (needed the USB port). The repository was only approx 1,3TB

**Comment by RonnyPfannschmidt** [comment]
I consider packs essential 

An alternative would be key value stores that optimize content addressing

**Comment by ThomasWaldmann** [comment]
@dietmargoldbeck what you've seen are 33MB/s.

That's not too bad for an initial backup to an USB drive. Initial backups always feel very slow just due to the amount of data and processing.

**Comment by tve** [comment]
Is this expected to break compatibility with the current (pre-pack) 2.0 betas?

**Comment by ThomasWaldmann** [comment]
@tve that likely will be the case.

in general, there is no compatibility assertion between betas and no upgrade code for beta-to-beta upgrades is provided.

**Comment by cr1901** [comment]
I run `borg` (v1 implied) on a `raidz`<sup>1</sup> array. I've been doing experiments with `borg-2`, and notice that archives I create with `borg-2` been taking more space than the equivalent `borg` archives. I finally [figured it out](https://www.truenas.com/community/threads/the-problem-with-raidz.114529/).

First off, according to my notes, `borg-2` does a fine job deduplicating; nothing I describe below can really be contributed to `borg-2` deduplication. See the Apparent Disk Usage column for an example repo I backed up with `borg` vs `borg-2` (usage is bytes):

|Borg Version |Real Disk Usage|Apparent Disk Usage|Dirs   |Files      |
|---------------|-----------------|-----------------------|------|---------|
|v2                  |30390315008   |24509081801            |65797|681435 |
|v1                  |23415717376   |23456004666            |4        |287       |

## `raidz` Overhead

The actual disk usage (`du -sB1`) above differs dramatically from apparent disk usage (`du -sb`). The TLDR is, for a `raidz` system, if many files on your filesystem have _both_ the following properties:

1. Small relative to `recordsize` (usually `128k` bytes, `1M` bytes is sometimes common).
2. Not close to a multiple of 8kB (2 sectors of a 4k bytes/sector disk).

You will have significant space overhead actually storing these files on your array due to how `raidz` works. In the above case, this is an overhead of 25%<sup>2</sup>!

```
(30390315008 - (1000 * 1000 * 1000)) / (24509081801 - (1000 * 1000 * 1000)) = 1.2502
```

I subtracted 1G because my scripts run `borg repo-space --reserve 1G` beforehand. This is with the default chunker (or close enough, I may have changed the window size to `16383`, but nothing else). With `--chunker-params buzhash,10,23,16,4095`, I saw even larger overheads:

```
Real: 28227564544
Apparent: 20624082512
Overhead: 1.387
```

This overhead [gets worse](https://www.truenas.com/community/threads/is-raid10-mirrored-vdevs-still-the-disk-layout-to-choose-at-all-flahs-storage-servers-for-esxi.93629/#post-648044) for `raidz2` and `raidz3`, but the TLDR above re: 8kB seems to still hold.

## Why Pack Files Are Useful For `raidz`

This overhead issue does not exist in `borg` because `borg` creates files around 500 MB in size. In my 2.18TB repo w/ around ~4000 of these files, and ZFS `-o recordsize=1M -o compress=off`, I'm losing ~4GB. This is all due to not enabling compression (see mitigations); the `raidz` overhead basically doesn't matter at this point. And even this overhead isn't worth me starting over:

```
100 * (4 / (2.18 * 1000)) = 0.18
```

Not even 1/5th of a percent<sup>3</sup>! If `borg-2` supported pack files, the small file overhead could possibly be amortized.


## Data

Here are some numbers on _another_ `borg-2` repo, _separate_ from the above one<sup>4</sup>:

### Size (Full)

```
Real: 24297848832
Apparent: 20723652906
Overhead: 1.181
```

Overhead once again subtracts 1GB from apparent and real.

### File Size Distribution (Full)

```
[32, 64): 0
[64, 128): 0
[128, 256): 9685
[256, 512): 54204
[512, 1024): 39688
[1024, 2048): 30403
[2048, 4096): 36669
[4096, 8192): 34776
[8192, 16384): 20283
[16384, 32768): 12367
[32768, 65536): 8879
[65536, 131072): 6908
[131072, 262144): 4079
[262144, 524288): 2388
[524288, 1048576): 2765
[1048576, 2097152): 2433
[2097152, 4194304): 1763
[4194304, 8388608): 786
[8388608, 16777216): 98
Total: 268174
```

A _lot_ of files have sizes < 8192, no doubt due to `borg-2` compression before encryption. This shows up as the 18% overhead. Perhaps there's a formula that can _approximate_ the overhead without having to run `du` multiple times, but I haven't come up with it yet. Of course `raidz2` and `raidz3` overheads would need different calculations.<sup>1</sup>

### Size (Single)

```
Real: 24297848832
Apparent: 20723652906
Overhead: 1.195
```

Note the overhead for any single backup tends to be worse than the overhead for many backups of the same tree (though I _have_ seen adding a single backup make the full overhead worse).

### File Size Distribution (Single)

```
[32, 64): 0
[64, 128): 0
[128, 256): 9232
[256, 512): 53065
[512, 1024): 37997
[1024, 2048): 27980
[2048, 4096): 31550
[4096, 8192): 21587
[8192, 16384): 12887
[16384, 32768): 8827
[32768, 65536): 4553
[65536, 131072): 3249
[131072, 262144): 1808
[262144, 524288): 1404
[524288, 1048576): 2267
[1048576, 2097152): 2141
[2097152, 4194304): 1696
[4194304, 8388608): 778
[8388608, 16777216): 98
Total: 221119
```

One backup tends to distribute  `borg-2` files sizes similarly for multiple backups. Therefore, in the future, I may streamline data collection by assuming a single backup generates a histogram whose bins have nearly the same height relative to one another.

## Mitigations

* One thing that can help- enable `zfs set compress=on`. Yes, even with `borg` supplying its own compression, this _does_ help a bit because ZFS records must be a power of 2 between sector size and `recordsize`. In ZFS, files in size between powers of two are zero-padded to the next power of 2 _or_ next integer multiple of `recordsize`. That zero padding _does_ compress well. I inexplicably didn't keep my numbers for this, but **IIRC** I saw something like 4-5% less space usage for `recordsize=128k` (~32GB became 30.8GB, as above) and `recordsize=1M` (~32GB became 30.4GB, as above).
* Additionally my own _limited_ experience so far is that frequently and consistently backing up amortizes the cost of the overhead. I would have to have detailed reference-counting info to know why.
* _Unless `encryption=none`_, I don't believe disabling `borg-2` compression and relying only on ZFS compression will help you, since compression is done before the encryption. ZFS won't know how to compress anything further except the aforementioned zero padding.

I think that covers everything. If you decide not to implement packing at all, I think there should be info in the manual about tuning `borg-2` for `raidz` systems. Perhaps this reply could be used as a base :D!

## Footnotes

1. [`raidz`](https://en.wikipedia.org/wiki/Non-standard_RAID_levels#RAID-Z) is equivalent to `raidz1`; it is similar to [RAID5](https://en.wikipedia.org/wiki/Standard_RAID_levels#RAID_5) in that one drive can fail without compromising data. **All numbers/approximation are directly valid for `raidz` only**, as I don't have the hardware (or the $ :P) to test `raidz2` or `raidz3`. Maybe I can get someone I trust who _does_ to let me run a script to get some numbers.
2. This is not unique to `borg-2`. On a `raidz` system, try this:
    
   ```
   wget https://web.archive.org/web/20120105073610/http://marknelson.us/attachments/million-digit-challenge/AMillionRandomDigits.bin
   mkdir digits
   split -b4096 AMillionRandomDigits.bin digits/amrd
   du -sB1 AMillionRandomDigits.bin
   du -sb AMillionRandomDigits.bin
   du -sB1 digits/
   du -sb digits/
   ```

   I get the following `du`s:
   
   ```
   535040  AMillionRandomDigits.bin
   415241  AMillionRandomDigits.bin
   637952  digits/
   415345  digits/
   ```

3. The actual overhead is closer to 0.25%, but I don't have the actual numbers handy right now.
4. It took me a few nights to get a system down for collecting data.

**Comment by SimonPilkington** [comment]
The new format also significantly increases `borg check` time, at least on a btrfs-formatted HDD. The drive very audibly seeks throughout while checking the borg2 repository but is almost silent with the borg1 repository. 

(Repositories are created from the same dataset.)

```
$ time borg check --repository-only borg1/

real    4m36.262s
user    0m13.271s
sys     0m26.063s
```

```
$ time borg2 check --repository-only -r borg2/

real    18m39.614s
user    0m32.492s
sys     0m30.501s
```

**Comment by akostadinov** [comment]
IMHO simplicity is something that has value on its own and if reverse chunking is implemented within Borg, there is no way out of this complexity.

Doesn't it make more sense to have reverse chunking as a separate optional layer instead of building this into Borg? Why don't we let filesystem optimizations and standalone backends deal with that (optionally) instead of complicating the system for everybody?

There are already filesystems that combine smaller files into the same block. Also there are requests for rclone and fuse backends to do reverse chunking of files (not implemented). My vote goes to keep things simple and resilient. And whoever needs the big files approach, use a standalone layer with it's pros and cons.

As an example one may choose not to use chinking for the madia library but use it for OS files and user homes. Or somebody wont care for the space overhead and want things simple.

Or if it has to live within borg, maybe do it as a re option - to use index or not. That index might be recreated if corrupt from the bundles' metadata.

**Comment by ThomasWaldmann** [comment]
I would very much like to let the storage deal with storage space management, that's what they are made for. The current repository storage is beautifully simple.

But the big problem is the amount of api calls multiplied by their latency and we can't get around that if we store many very small objects.

**Comment by cr1901** [comment]
Here is the Python script I use to generate histogram values (slightly modified, modifications untested). Maybe it's worth running and/or adding to `borg-2` as a diagnostic tool to figure out what the typical distribution of chunks looks like?

```python
#!/usr/bin/env python3

import subprocess
import os
from tqdm import tqdm
from itertools import pairwise

MEGA=1024*1024
KILO=1024

ranges = map(lambda exp: int(KILO*(2**exp)), range(-5,15))

ROOT = os.environ.get("BORG_REPO")

hist = {}

for (start, end) in tqdm(pairwise(ranges)):
    hist[(start, end)] = int(subprocess.run("find {ROOT}/data -type f -size +{start - 1}c -a -size -{end}c | wc -l",
                                        shell=True, check=True, capture_output=True).stdout)

for k,v in hist.items():
    print(f"[{k[0]}, {k[1]}): {v}")

print(f"Total: {sum(hist.values())}")
```

**Comment by ThomasWaldmann** [comment]
@cr1901 i did something similar a while ago when working on the chunkers, even with graphical charts, but there was nothing special to see. the chunk size distribution is not the problem, it is whether we store chunks separately (like borg2 now) or many of them packed together in a big file (somehow like borg 1.x, but assembled client-side).

**Comment by SimonPilkington** [comment]
> Doesn't it make more sense to have reverse chunking as a separate optional layer instead of building this into Borg? Why don't we let filesystem optimizations and standalone backends deal with that (optionally) instead of complicating the system for everybody?
> 
> There are already filesystems that combine smaller files into the same block. Also there are requests for rclone and fuse backends to do reverse chunking of files (not implemented). My vote goes to keep things simple and resilient. And whoever needs the big files approach, use a standalone layer with it's pros and cons.

@akostadinov do you have any concrete recommendations? If borg isn't going to perform packing then it would be nice to have information on what works well and what doesn't. 

btrfs on an HDD evidently performs poorly, per my previous message.

**Comment by PhrozenByte** [comment]
tl;dr

I strongly vote for packing.

A few suggestions for a possible implementation can be found at the end.

## About issues

> IMHO simplicity is something that has value on its own

I agree, in principle, but in this case simplicity comes at a cost.

#### Storage overhead and metadata limitations

Take [William's findings](https://github.com/borgbackup/borg/issues/8572#issuecomment-3344805895). Sure, you can argue that the storage overhead created is an issue of ZFS' raidz implementation, but this isn't a problem limited to ZFS: Basically all filesystems (fs) perform poorly with many small files. Yes, many modern fs have solutions to reduce the storage overhead, but other fs suffer the same or other limitations.

For example, ext4 has a limited number of inodes. ext4 creates 1 inode per 16 KiB by default. Depending on the fs type (`mkfs.ext4 -t fs-type`, see `/etc/mke2fs.conf`) it might create even less: 1 inode per 32 KiB for `big`, and 1 inode per 64 KiB for `huge` fs. Some distributions even create just 1 inode per 256 KiB, depending on the disk size and usage scenario (e.g., NAS distributions). One can't change this ratio after fs creation. Thus, every file below that treshold "eats" up inodes. One can easily run into a "storage full" situation without the disk actually being full. ext4 isn't alone here, e.g., XFS behaves very similar.

Such situations can also arise with different fs following vastly different approaches, e.g., Copy-on-Write (CoW) fs like btrfs. The "no space left on device" error has become infamous for btrfs, even though it's very rare in normal operation since a very long time. However, creating many small files is one of the "non-normal" situations which can cause a btrfs to eat up all its metadata-reserved space, especially when combined with snapshots, or multi-disk volumes.

#### I/O performance

Another issue with many small files all fs face is I/O performance, as pointed out by Thomas before. The reason is simple: Opening a file descriptor and issuing a read syscall is cheap on modern hardware, but opening millions of file descriptors and issuing millions of read syscalls is not ("even small things can make a mess"). This is true even on local NVMe storage, but it gets way worse with HDDs, and might get unbearable if we're not talking about local syscalls, but remote backends whose API calls need to go over the wire. Some remote protocols are extremely prone to slowdowns due to protocol overhead (e.g., SFTP, or WebDAV, or S3). Even protocols that were specifically designed for network filesystems like Samba/CIFS have infamous slowdowns when working with many small files. This can get as bad as that one can transfer magnitudes of more data in the same time if it's just packed, up to the point that the increased (network) I/O Thomas mentioned earlier is not just compensated, but transfers are still significantly faster.

Take the following SFTP example (~33 MBit/s uplink, ~25 ms RTT to the server, i.e., expecting ~25.4 seconds to transfer 100 MiB plus protocol overhead; SFTP requires 3 RTTs per file AFAIK, plus some generic protocol and encryption overhead; results below are actually the minimums of 10 runs each):

```console
$ sudo mount -t tmpfs -o size=1G --mkdir tmpfs ./sftp-bench
$ cd ./sftp-bench
$ mkdir single-big some-medium many-small
$ dd if=/dev/urandom of=single-big/file.img bs=100K count=1024 status=none                                               # create 1x 100 MiB file
$ for i in $(seq 0 49); do dd if=/dev/urandom of=some-medium/$(printf '%02d' $i).img bs=2M count=1 status=none; done     # create 50x 2 MiB files
$ for i in $(seq 1 1024); do dd if=/dev/urandom of=many-small/$(printf '%04d' $i).img bs=100K count=1 status=none; done  # create 1024x 100 KiB files
$ time scp -q -r single-big/ hetzner-storagebox:single-big
real	0m25,600s
user	0m0,205s
sys	0m0,212s
$ time scp -q -r some-medium/ hetzner-storagebox:some-medium
real	0m45,561s
user	0m0,201s
sys	0m0,227s
$ time scp -q -r many-small/ hetzner-storagebox:many-small
real	2m10,344s
user	0m0,377s
sys	0m0,669s
```

#### Fragmentation

On HDDs the problem gets significantly worse in high disk usage scenarios due to fragmentation. Packing has the inherent disadvantage of requiring `compact` to read the pack, drop unused entries and write it back to the store creating overhead. However, this is actually an advantage on HDDs, because it reduces fragmentation: It allows the HDD controller to free continuous space on the disk and thus allows Borg to read and write new chunks without too much seeking. Without packing `compact` doesn't necessarily free continuous space, but rather creates thousands of tiny free space holes, which Borg then fills up again with completely unrelated chunks in later runs. This causes fragmentation hell and causes a significant slowdown. Fs like btrfs and ZFS amplify this significantly due to their CoW approach - which is exactly what [Simon noticed](https://github.com/borgbackup/borg/issues/8572#issuecomment-3600340376) earlier.

But this gets even worse when copying an existing repo to another HDD-backed fs: The fs then looses any information about how data relates to other data. The fs only knows paths, so they store data with similar paths next to each other. For a Borg 2 repo this means that data is stored together in accordance to chunk ID similarity. That's basically the worst that can happen, because chunk IDs have literally no accordance to what data they store - on purpose! However, it means that the HDD will have to seek for basically every single chunk Borg reads. Defrag doesn't help at all - it actually makes things worse, because defrag also moves data according to paths.

#### Conclusion

Just to throw in a few more numbers: My personal backup repo consists of 3,628,766 unique chunks, the repo of my father consists of 594,760 unique chunks, the repo of my personal (but not purely private) server consists of 1,488,960 unique chunks, and I'm not even talking about the business-related stuff, which is way bigger.

So, to summarise, this is nothing one can circumvent by choosing the "right" storage backend - all storage backends have problems with many small files, it's just a matter of how bad things can get in practice.

## About 3rd-party solutions

> There are already filesystems that combine smaller files into the same block. Also there are requests for rclone and fuse backends to do reverse chunking of files (not implemented).

Besides such solutions not existing yet (AFAIK) and me having some doubts that they will, they neither solve the issue, nor do they have any advantage, even if such projects actually materialise. You argue about complexity, but moving complexity into a separate project (i.e., to rclone, or a FUSE fs) doesn't make that complexity disappear. It's still there.

Even worse, since separate projects don't have an understanding of what Borg is storing, the implementation doesn't know what to pack together, yielding bad results (see the paragraph about fragmentation on HDDs above). If Borg wants to pack chunks, it needs to do it on its own. Only Borg understands what chunks "belong to each other" and are more likely to be accessed together. This is an optimisation separate projects simply can't do, making a possible advantage of transferring bigger packs (see SFTP example above) go away completely, because way too much unrequested data is transferred as well.

Besides, FUSE is an optional dependency for Borg right now. Making it an requirement is a disadvantage on its own, especially because FUSE isn't always available. For example, Borg aims to add native Windows support (i.e., without WSL). But not just that, I usually run Borg in containers, which don't have FUSE support by default, but require additional care, setup, and allowing FUSE reduces overall system security.

I'm afraid there are no 3rd-party solutions to this problem.

## About Borg

I thus agree [with Ronny](https://github.com/borgbackup/borg/issues/8572#issuecomment-2514315356) a year back and consider packing essential.

IMHO it's more of a question how to implement it - and Thomas' suggestion to do client-side packing instead seems reasonable to me, especially because Borg 2 aims to support "dumb" storage backends like rclone, i.e., storages without a server-side process. However, packing is only reasonable if an algorithm can reasonably predict what chunks will be requested together. Packing by "transaction" like Borg 1 does is probably the most basic algorithm, but likely allows for further optimisation. How much more optimisation is possible in practice is something Thomas can assess best.

#### Suggestions for an implementation

To assess the "proximity" of chunks (i.e., the likelihood that chunks are requested together), Borg could use path-based information about chunk usage in different archives: if archives share many chunks below the same path, Borg can reasonably assume that new chunks below that path will be requested together with the older ones, thus adding new chunks to packs that already store the other chunks. If Borg has requested multiple packs, it shouldn't limit its operation to adding chunks to existing packs, but also to split up packs to allow for later additions. Since such checks are memory intensive, the user should be able to configure how "smart" Borg is allowed to be (i.e., how much memory it is allowed to use). Thus, Borg could limit proximity checks to archive series.

From theory I'd expect a size-based packer (i.e., not based on the amount of chunks per pack, but the overall size of all chunks in the pack) to perform better in most situations. In general, chunks of more or less the same size should be packed together, because packing a big chunk with many small chunks bears the risk of reading many big unrequested chunks just to get a few small ones. Chunks above some treshold shouldn't be packed at all, and that treshold should be significantly lower than the "target pack size". Since Borg must be able to add new chunks to existing packs, there should be a lower target pack size that is used when assembling new packs, and an upper target pack size that is used as upper limit when adding new chunks to an existing pack.

We don't need Borg 2 to create as few packs as Borg 1 did, just significantly fewer than without packing. From theory I'd expect that packs bigger than a very few MiB (likely single digit) don't have any advantage, even with high overhead protocols like SFTP. To reduce fragmentation on HDDs it might be reasonable not to use purely random filenames for packs, but use increments to allow fs to physically store packs that are expected to be requested together closer to each other.

Thus, it might be reasonable to implement a configurable packer that packs chunks in accordance to the storage backend used. For example, with a SFTP-based backend Borg should create bigger packs than with any local storage, because SFTP's protocol overhead is crazy bad and transmitting more unnecessary data can still be beneficial over all (see SFTP example above; WebDAV should be similar in this regard). Another example is S3-based storage: Some cloud providers bill by storage usage *and* object count, so creating even bigger packs than the storage backend would suggest (should be similar to SFTP) could be beneficial for monetary reasons. It might be best to have reasonable defaults per storage backend, but to allow the user to change these defaults. The only way to find good defaults is by real-world testing I assume.

However, making the packer configurable also means that Borg must be able to repack a repo (which could be combined with a `check`) if the repo is moved to a different storage backend (e.g., `borg2 repo-pack`, similar to `repo-compress`). This would also allow users to disable packing altogether - which can undoubtedly be best under specific circumstances, e.g., on an fast ext4 formatted NVMe with a low inode ratio.

#### Final thoughts

So far my analysis and suggestions on the matter. 

I'm well aware that I didn't make the implementation of a packer simpler, quite the contrary, but I strongly believe that it is worth the effort, because managing millions of files - i.e., a repo I consider rather mid sized - will cause problems for sure. But maybe it is a nice technical challenge to create the most efficient solution 😊

**Comment by akostadinov** [comment]
I'm not the one to implement this and don't know the architecture of borg so I'm just saying that client side seems to me like the wrong layer the reverse chunking should be done on.

From a separation of concerns standpoint of view, that seems to me to be somewhere more on the backend. Because different backends may need different optimizations or not need at all.

I understand that some backends will work very bad without reverse chunking (it is not actually chunking because this already exists in rclone and is to limit how big a file can be).

But one needs to keep in mind that if reverse chunking is implemented in the application (borg) it will eventually hit the same issues that the backend hits, trying to implement these. For example btrfs had some ideas to implement block suballocation but that never went off I assume for complexity reasons. btw [ext4](https://docs.kernel.org/next/filesystems/ext4/inlinedata.html) and [btrfs](https://btrfs.readthedocs.io/en/latest/Inline-files.html) already support inlining very small files in the inodes.

The only reason I see that such reverse chunking would work in Borg but not generally on the FS level is that an FS has to provide generic performance while it might be easier to schedule some maintenance/defragmentation/cleanup work on backups.

But then why not implement the reverse chunking for simplicity on a middle, still standalone layer like rclone or fuse? For media file backup you would not need reverse chunking like you would need on your user home dir. And I (probably many others) have some limited amount of such files to backup so maybe they don't care about a few lost GB. But care about simple storage that doesn't break and is easy to navigate.

Also such an independent reverse chunking layer may have it's own life as a project used and improved by many, thus reducing maintenance cost and improving quality.

What is more concerning is remote storage to S3 and other object storage. It will probably benefit from chunking be it on the app level or on a middle layer. But I observe that software dealing with remote data like package managers, container image registries, torrents, etc. all tend to benefit huge from 5+ concurrent operations. While for a local HDD sequential operations are much preferred.

So IMHO it is much more important to build into the system ability to parallelize operations via threads or some async architecture (maybe Borg already does that), which will allow to configure sequential or parallel operations depending on the underlying storage. This cannot transparently be done at an intermediate layer.

So I'm not against having reverse chunking, I'm all for it, just saying it may make more sense to be on its own layer. Parallelism - a must.

**Comment by tve** [comment]
I decided to try out borg 2 this evening (long time borg 1 user) and I revisited this discussion. Ran some tests using rclone copy of a 1GB file, vs. 500 2MB files vs. 10K 100KB files. Pushed them to Hetzner S3 using concurrency of 16/64/128 for the small files. Server is 10ms away from Hetzner endpoint. I can't get the small files to go more than 1/10th the speed of the medium files and there clearly are plenty of hiccups and even errors thrown by Hetzner.

The tried a borg backup of a filesystem with 7GB of files and it's another one or two orders of magnitudes slower than any of the rclone copy tests.

Not sure what to make of this. I thought of running the tests against AWS S3 from an EC2 instance to see what the realistic best case of using small objects is. But maybe I should first ask: is there an open question in terms of what the magnitude of the trade-off is?

**Comment by tve** [comment]
Following up on my previous post: I decided to do another experiment, which is to run borg 2.0.0v20 over ZeroFS (https://www.zerofs.net/) on the same Hetzner bucket as in my previous test. The difference is like day and night: using the borg rclone back-end it took several hours to get less than 3GB backed-up and it ended up aborting on some Hetzner error. With ZeroFS the full 9GB backup completed in 16 minutes, plus around 10 minutes for ZeroFS to finish compactions. This equates to 9MB/s or 6MB/s of upload speed depending on whether one counts the ZeroFS compactions time or not.

Some thoughts:
- the 6MB/s-9MB/s range is about the same speed I see running a little rclone benchmark with 100KB files and significant parallelism (64 or 128 concurrency), it's 1/10th-1/5th the speed of the same rclone benchmark with 2MB files
- this test shows that it is possible to get reasonable performance given the Hetzner back-end
- the ZeroFS logs showed a fair share of low-level Hetzner errors, when I tried it a few months ago one of these would eventually cause the filesystem to abort, lots of work has gone in to it in the meantime to make it resilient
- the ZeroFS architecture is completely incompatible with the borg implementation and philosophy: I'm not suggesting that this is _the_ solution, however it works and provides a proof of existence and a performance datapoint
- overall it seems to me that there are two performance gaps in borg: one is in producing sufficient parallelism to saturate the storage back-end, the second is in generating large enough objects to use the storage back-end within an optimal operating regime. It's not clear to me at all which of these two gaps is bigger at the moment.

NB: I apologize that my results are all of the handwavy kind. I was fooling around to get a grip on orders of magnitude and on what's possible. I can refine my scripts if there's interest.

**Comment by Ogak-AI** [comment]
Hi Everyone, my name is Finomo Awajiogak. I am currently a student of NIIT Port Harcourt here in Nigeria. I am glad to apply as a contributor for Borg Collective for Google Summer of Code 2026. I would be glad to work on this issue.

**Comment by ThomasWaldmann** [comment]
@Ogak-AI Yes, this is one of our GSOC 2026 project ideas from the ideas list. If you want apply for working on this, file a good project application via google's gsoc site.

**Comment by ThomasWaldmann** [comment]
Congrats to @mr-raj12 for doing this as a GSOC 2026 project!

**Comment by ThomasWaldmann** [comment]
# Packs sneak peek
- the focus of this benchmark is how **latency** impacts borg operations
- extract is not optimized yet
- in general packs work is not finished yet
- check: sha256 check of pack is done server-side when using borg serve --rest
- Run 1: single-object packs, latency- and overhead-wise quite similar to what we had before the packs project of @mr-raj12 
- Run 2: packs with ~50MB size (not as big as the 500MB borg 1.x used)

# rest:// PackWriter performance test

Effect of `PackWriter` packing strategy on borg2 performance over a `rest://`
repository, under emulated network latency. Two full runs were made, differing only
in how chunks are grouped into pack files.

## Test parameters

| | |
|---|---|
| borg version | `2.0.0b23.dev201+g8addc5841` |
| Repository | `rest://tw@localhost//tmp/REPO` (served by `borg serve --rest`, FILE backend, over localhost sshd) |
| Encryption | `none` (`borg repo-create --encryption none`) |
| Harness | `scripts/rest_perf_test.sh` |
| Date | 2026-06-30 |

### Test data (incompressible random, generated once, identical across all runs)

| group | count | size each | total |
|---|--:|--:|--:|
| tiny | 20,000 | 20 KiB | ~400 MiB |
| med | 1,000 | 5 MiB | ~5 GiB |
| big | 2 | 1 GiB | 2 GiB |
| **total** | **21,002 files** | | **~7.4 GiB (orig_size 7.8 GB)** |

### Network profiles

Latency/bandwidth are emulated client-side by borgstore
(`BORGSTORE_LATENCY` µs, `BORGSTORE_BANDWIDTH` bit/s). **LAN and WAN share the same
1 Gbit/s bandwidth and differ only in latency**, to isolate the effect of latency.

| profile | latency | bandwidth |
|---|--:|--:|
| raw | 0 (off) | unlimited (off) |
| lan | 0.3 ms | 1 Gbit/s |
| wan | 30 ms | 1 Gbit/s |

### Workload (per profile)

1. `repo-create --encryption none`
2. **backup1** — `create` of the full data set (cold)
3. **backup2** — `create` of the same data again (full dedup)
4. **extract-n** — `extract --dry-run` of backup2
5. **check** — `check` of the repository

### PackWriter variants

The single `PackWriter(...)` instantiation in `src/borg/repository.py` was changed per run
(restored to the default `max_count=3` afterwards):

| run | configuration | effect |
|---|---|---|
| **Run 1** | `max_count=1, max_size=None` | one chunk per pack → ~24k tiny pack objects |
| **Run 2** | `max_count=None, max_size=50000000` | 50 MB packs → ~150 pack objects |

All values are wall-clock seconds, one column per network profile (`raw` / `lan` / `wan`).
`original_size` was a constant 7.800 GB across every step and profile.

## Run 1 — `max_count=1, max_size=None`

| step | raw | lan | wan |
|---|--:|--:|--:|
| backup1 | 131.746 | 150.051 | 1047.596 |
| backup2 (dedup) | 15.461 | 16.771 | 16.368 |
| extract-n | 33.092 | 83.626 | 982.148 |
| check | 25.702 | 21.409 | 914.071 |

## Run 2 — `max_count=None, max_size=50000000`

| step | raw | lan | wan |
|---|--:|--:|--:|
| backup1 | 52.891 | 90.751 | 95.861 |
| backup2 (dedup) | 15.468 | 15.467 | 16.063 |
| extract-n | 32.096 | 98.197 | 989.980 |
| check | 5.072 | 5.336 | 14.493 |

## Direct comparison (wall_s, Run 1 → Run 2)

| profile | step | Run 1 | Run 2 | speedup |
|---|---|--:|--:|--:|
| wan | backup1 | 1047.6 | 95.9 | **10.9×** |
| wan | extract-n | 982.1 | 990.0 | 0.99× (no change) |
| wan | check | 914.1 | 14.5 | **63.1×** |
| raw | backup1 | 131.7 | 52.9 | 2.5× |
| raw | check | 25.7 | 5.1 | 5.1× |

## Findings

- **Large packs are decisive for write and check over high latency.** With `max_count=1`,
  each chunk is a separate `store()`/object, so WAN `backup1` and `check` pay ~30 ms ×
  tens of thousands of calls. 50 MB packs cut backend calls to ~150: backup1 **11× faster**,
  check **63× faster** on WAN.
- **Packing does not help `extract`.** WAN `extract --dry-run` is ~982–990 s in *both* runs.
  borg's read path (`Archive.fetch_many` → `Repository.get_many` → `get`) issues **one
  range-read `store.load` per chunk**, serially, with no pack-level grouping or cache
  (borgstore cache is `C_OFF` for `packs/`). So extract latency tracks *chunk count*, not
  *pack count*. A pack-cache (borgstore `C_WRITETHROUGH` on `packs/`) or a pack-aware
  `get_many` would be needed to benefit WAN restores — and only with large packs.
- **Tiny packs cost even locally.** `raw` (no emulation) backup1 is 2.5× slower and check
  5× slower with `max_count=1`, from the filesystem overhead of ~24k small pack files.
- **backup2 (full dedup) is ~15–16 s everywhere** — almost no `store()` calls, so neither
  latency nor pack size matters; a useful sanity baseline.

**Comment by ThomasWaldmann** [comment]
TODO: close after b22 is released with packs.

**Comment by ThomasWaldmann** [comment]
2.0.0b22 soon (waiting for CI, binaries).
